android-kotlin/app/src/main/java/com/mongodb/app/ItemsActivity.kt

```diff
@@ -1,39 +1,154 @@
+
+        fab = findViewById(R.id.floating_action_button)
+        fab.isEnabled = true
+        fab.setOnClickListener { (addItem()) }
+
+        recyclerView = findViewById(R.id.items_list)
+
+        val toolbar = findViewById<View>(R.id.task_menu) as Toolbar
+        setSupportActionBar(toolbar)
+        toolbar.showOverflowMenu()
+    }
+
+    // Ensure the user realm closes when the activity ends.
+    override fun onDestroy() {
+        super.onDestroy()
+        userRealm?.close()
+        recyclerView.adapter = null
+    }
+    override fun onStop() {
+        super.onStop()
+        realmApp.currentUser().run {
+            userRealm?.close()
+        }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu): Boolean {
+        menuInflater.inflate(R.menu.task_menu, menu)
+        return true
+    }
+
+    /**
+     *  Logs out the user and brings them back to the login screen.
+     */
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        val user = realmApp.currentUser()
+        return when (item.itemId) {
+            R.id.action_logout -> {
+                user?.logOutAsync {
+                    if (it.isSuccess) {
+                        realmApp.removeUser(realmApp.currentUser())
+                        Log.v(TAG(), "user logged out")
+                        startActivity(Intent(this, LoginActivity::class.java))
+                    } else {
+                        Log.e(TAG(), "log out failed! Error: ${it.error}")
+                    }
+                }
+                true
+            }
+            else -> {
+                super.onOptionsItemSelected(item)
             }
         }
     }
+
+    /**
+     *  Allows the user to insert an item into the realm.
+     */
+    private fun addItem() {
+        // Create a dialog to enter an item name.
+        val input = EditText(this)
+        val dialogBuilder = AlertDialog.Builder(this)
+        dialogBuilder.setMessage("Enter item name:")
+                .setCancelable(true)
+                .setPositiveButton("Create") { dialog, _ ->
+                    run {
+                        dialog.dismiss()
+                        // Add a new item to the list by inserting into the realm.
+                        val item = Item(input.text.toString())
+                        // All realm writes need to occur inside of a transaction.
+                        userRealm?.executeTransactionAsync { realm ->
+                            realm.insert(item)
+                        }
+                    }
+                }
+                .setNegativeButton("Cancel") { dialog, _ ->
+                    dialog.cancel()
+                }
+        val dialog = dialogBuilder.create()
+        dialog.setView(input)
+        dialog.setTitle("Add Item")
+        dialog.show()
+        input.requestFocus()
+    }
+
+    private fun setUpRecyclerView(realm: Realm, user: User?, partition: String) {
```

Comment:
This API feels unintentional.

- We have the realm is the userRealm member
- We only call this when we know the user is logged in, so the user doesn't have to be optional. Besides, we have global access to the realmApp and can always get the currentUser() from that.
- The partition should be readable from the realm itself (not sure on Kotlin, but it is in Swift). Besides, the partition value doesn't seem to be used by the adapter?

I would recommend rethinking this method a bit.

Reply:
I decided to inline this functionality into the onStart method, let me know your thoughts.